Mage/src/main/java/mage/abilities/effects/common/DoIfCostPaid.java

```diff
@@ -100,69 +100,64 @@ public boolean apply(Game game, Ability source) {
             boolean result = true;
             Outcome payOutcome = executingEffects.getOutcome(source, this.outcome);
-            if (cost.canPay(source, source, player.getId(), game)
-                    && (!optional || player.chooseUse(payOutcome, message, source, game))) {
+            boolean canPay = cost.canPay(source, source, player.getId(), game);
+            if (canPay && (!optional || player.chooseUse(payOutcome, message, source, game))) {
                 cost.clearPaid();
                 int bookmark = game.bookmarkState();
                 if (cost.pay(source, game, source, player.getId(), false)) {
                     game.informPlayers(player.getLogName() + " paid for " + mageObject.getLogName() + " - " + message);
-                    if (!executingEffects.isEmpty()) {
-                        for (Effect effect : executingEffects) {
-                            effect.setTargetPointer(this.targetPointer);
-                            if (effect instanceof OneShotEffect) {
-                                result &= effect.apply(game, source);
-                            } else {
-                                game.addEffect((ContinuousEffect) effect, source);
-                            }
-                        }
-                    }
+                    result &= applyEffects(game, source, executingEffects);
                     player.resetStoredBookmark(game); // otherwise you can e.g. undo card drawn with Mentor of the Meek
                 } else {
+                    game.informPlayers(player.getLogName() + " did not pay for " + mageObject.getLogName() + " - " + message);
                     // Paying cost was cancels so try to undo payment so far
                     player.restoreState(bookmark, DoIfCostPaid.class.getName(), game);
-                    if (!otherwiseEffects.isEmpty()) {
-                        for (Effect effect : otherwiseEffects) {
-                            effect.setTargetPointer(this.targetPointer);
-                            if (effect instanceof OneShotEffect) {
-                                result &= effect.apply(game, source);
-                            } else {
-                                game.addEffect((ContinuousEffect) effect, source);
-                            }
-                        }
-                    }
+                    result &= applyEffects(game, source, otherwiseEffects);
                 }
-            } else if (!otherwiseEffects.isEmpty()) {
-                for (Effect effect : otherwiseEffects) {
-                    effect.setTargetPointer(this.targetPointer);
-                    if (effect instanceof OneShotEffect) {
-                        result &= effect.apply(game, source);
-                    } else {
-                        game.addEffect((ContinuousEffect) effect, source);
-                    }
+            } else {
+                if (canPay) {
```

Comment:
I think this leaks unnecessary info? Some costs use hidden information e.g. reveal a card from your hand. So better to use the "did not" inform text in all cases.

Reply:
Oh right. Forgot about costs from hidden zones.